File: leetcode/0004/python/main.py
import unittest

class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """

        assert nums1 or nums2

        n = len(nums1)
        m = len(nums2)

        if n > m:
            return self.findMedianSortedArrays(nums2, nums1)

        half = (n + m)//2

        low = 0
        high = n
        while low <= high:
            x = (low + high)//2
            y = half - x

            if x > 0 and nums1[x-1] > nums2[y]:
                high = x - 1
                continue
            if x < n and nums2[y-1] > nums1[x]:
                low = x + 1
                continue
            break

        if x == n:
            t2 = nums2[y]
        elif y == m:
            t2 = nums1[x]
        else:
            t2 = min(nums1[x], nums2[y])

        if (n + m) % 2: # odd
            return t2

        if x == 0:
            t1 = nums2[y-1]
        elif y == 0:
            t1 = nums1[x-1]
        else:
            t1 = max(nums1[x-1], nums2[y-1])

        return (t1 + t2)/2

class TestSolution(unittest.TestCase):

    def setUp(self):
        self.solution = Solution()

    def test_case1(self):
        nums1 = [1, 3]
        nums2 = [2]
        res = self.solution.findMedianSortedArrays(nums1, nums2) 
        self.assertEqual(res, 2)

    def test_case2(self):
        nums1 = [1, 2]
        nums2 = [3, 4]
        res = self.solution.findMedianSortedArrays(nums1, nums2) 
        self.assertEqual(res, 2.5)

    def test_case3(self):
        nums1 = [0, 0]
        nums2 = [0, 0]
        res = self.solution.findMedianSortedArrays(nums1, nums2) 
        self.assertEqual(res, 0)

    def test_case4(self):
        nums1 = []
        nums2 = [1]
        res = self.solution.findMedianSortedArrays(nums1, nums2) 
        self.assertEqual(res, 1)

    def test_case5(self):
        nums1 = [2]
        nums2 = []
        res = self.solution.findMedianSortedArrays(nums1, nums2) 
        self.assertEqual(res, 2)

    def test_casey(self):
        nums1 = [3, 4]
        nums2 = []
        res = self.solution.findMedianSortedArrays(nums1, nums2) 
        self.assertEqual(res, 3.5)

if __name__ == '__main__':
    unittest.main()

File: leetcode/0004/cpp/main.cpp
#include <vector>
#include <algorithm>
#include <cassert>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        size_t n = nums1.size();
        size_t m = nums2.size();

        if (n > m)
            return findMedianSortedArrays(nums2, nums1);

        size_t half = (n + m)/2;
        size_t low = 0;
        size_t high = n;
        size_t x, y;
        while (low <= high) {
            x = (low + high)/2;
            y = half - x;

            if (x > 0 && nums1[x-1] > nums2[y])
                high = x - 1;
            else if (x < n && nums2[y-1] > nums1[x])
                low = x + 1;
            else
                break;
        }

        float t1, t2;
        if (x == n)
            t2 = nums2[y];
        else if (y == m)
            t2 = nums1[x];
        else
            t2 = std::min(nums1[x], nums2[y]);

        if ((n + m) & 1) // odd
            return t2;

        if (x == 0)
            t1 = nums2[y-1];
        else if (y == 0)
            t1 = nums1[x-1];
        else
            t1 = std::max(nums1[x-1], nums2[y-1]);

        return (t1+t2)/2.0f;
    }
};

int main() {
    Solution solution;
    {
        vector<int> nums1{1, 3};
        vector<int> nums2{2};
        assert(solution.findMedianSortedArrays(nums1, nums2) == 2);
    }
    {
        vector<int> nums1{1, 2};
        vector<int> nums2{3, 4};
        assert(solution.findMedianSortedArrays(nums1, nums2) == 2.5);
    }
    {
        vector<int> nums1{0, 0};
        vector<int> nums2{0, 0};
        assert(solution.findMedianSortedArrays(nums1, nums2) == 0.0);
    }
    {
        vector<int> nums1{1};
        vector<int> nums2;
        assert(solution.findMedianSortedArrays(nums1, nums2) == 1.0);
    }
    {
        vector<int> nums1{};
        vector<int> nums2{2};
        assert(solution.findMedianSortedArrays(nums1, nums2) == 2.0);
    }
}
